### ISSUE TITLE
t.i.p.BaseProtocol mentions unknow 'dataReceived'

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5131]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)| @multani reported|
|-|-|
|Trac ID|trac#5131|
|Type|defect|
|Created|2011-06-02 07:00:40Z|
|Branch|https://github.com/twisted/twisted/tree/baseprotocol-docstring-5131|

Reading the API documentation of t.i.p.BaseProtocol, it says that:

```
... My API is quite simple. Implement dataReceived(data) ...
```

But dataReceived() is actually implemented by the subclass in t.i.p.Protocol.

If you read the documentation of BaseProtocol only, there's no way (AFAIK) that dataReceived() implemented by your subclass is going to be called. So I think andthe docstring of BaseProtocol should be updated not to mention dataReceived(). Otherwise, it just confuses reader.

<details><summary>Searchable metadata</summary>

```
trac-id__5131 5131
type__defect defect
reporter__multani multani
priority__normal normal
milestone__ 
branch__branches_baseprotocol_docstring_5131 branches/baseprotocol-docstring-5131
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1306998040000000 1306998040000000
changetime__1310643540000000 1310643540000000
version__None None
owner__exarkun exarkun
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Well... It's true that `BaseProtocol` is shared between `Protocol`, `ProcessProtocol`, and `DatagramProtocol`.  As such, the mention of the `dataReceived` method is somewhat out of place.

I'm not sure _If you read the documentation of BaseProtocol only, there's no way (AFAIK) that dataReceived() implemented by your subclass is going to be called._ makes sense, though.  `dataReceived` **is** called on `Protocol` instances that receive data.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_2"></a>|@multani commented|
|-|-|

Replying to [exarkun](#note_1):
> Well... It's true that `BaseProtocol` is shared between `Protocol`, `ProcessProtocol`, and `DatagramProtocol`.  As such, the mention of the `dataReceived` method is somewhat out of place.

Can't this be just moved to the `t.i.p.Protocol` docstring instead, and update `BaseProtocol` docstring to say to refer to subclass for the way of reading data out of the transport?

> I'm not sure _If you read the documentation of BaseProtocol only, there's no way (AFAIK) that dataReceived() implemented by your subclass is going to be called._ makes sense, though.  `dataReceived` **is** called on `Protocol` instances that receive data.

Actually, I was thinking that if you passed a `BaseProtocol` instance, `dataReceived()` would not be called at all, but I was wrong (this is the transport which calls this).
That's also mean that if you implement this method, there are great chances that you are going to use a transport which expect a `IProtocol`, which should also have `connectionLost()`, right?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [32262]) Branching to 'baseprotocol-docstring-5131'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Thanks for the additional feedback multani.  I've checked some doc changes into the branch.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_5"></a>|@multani commented|
|-|-|

I think this is much better than before: it removes the confusion from `BaseProtocol` and add more documentation to the (more directly useful) `Protocol` class.

I consider the original report to be fixed with those changes.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess set owner to @exarkun|
|-|-|

Thanks for working on this, exarkun. Looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [32336]) Merge baseprotocol-docstring-5131

Author: exarkun
Reviewer: multani, jesstess
Fixes: [#5131](https://github.com/twisted/twisted/issues/5131)

Give `Protocol` a class docstring, mostly taken from `BaseProtocol`'s docstring,
which was really describing a lot of things particular to `Protocol`.  Also give
`BaseProtocol` a docstring more appropriate to it.
